### PR TITLE
Container deployer private registry

### LINKED ISSUE #39: Support pull from private repositories in Container Deployer
**How to categorize this issue?**

/area security
/priority normal

**What would you like to be added**:

It should be possible to pull blueprints and component descriptors from a private repository in the container deployer.
Specifically it has to be enabled in the init container of the deployer where currently it is only possible to directly specify docker pull secrets.

### PR DESCRIPTION
**Pull request by hendrikKahl**
**How to categorize this PR?**

/area security
/kind enhancement
/priority normal

**What this PR does / why we need it**:
In order to pull artifacts from private registries, the container deployer syncs the respective secrets to the host cluster and makes them available to pod executing the deploy item.
Specifically, it mounts the OCI secrets for blueprint and component descriptor to the init-container and makes the imagePullSecret available to the pod in general.

The container deployer also implements the containerd default for images pulled from Dockerhub by adding `index.docker.io` to the repository string.

In addition, the helm chart exposes a parameter to pull from registries via http.

**Which issue(s) this PR fixes**:
Fixes #39 

**Special notes for your reviewer**:

**Release note**:

``` improvement user
The container deployer supports pulling OCI (blueprints, component descriptor, container images) artifacts from private registries by supplying respective auth credentials.
```

### CODE REVIEW

File: hack/install.sh [hunk 1]
```diff
@@ -13,7 +13,7 @@ if [[ $EFFECTIVE_VERSION == "" ]]; then
   EFFECTIVE_VERSION=$(cat $PROJECT_ROOT/VERSION)
 fi
 
-CGO_ENABLED=0 GOOS=linux GOARCH=amd64 GO111MODULE=on \
+CGO_ENABLED=0 GOOS=$(go env GOOS) GOARCH=amd64 GO111MODULE=on \
```
Comment (schrodit) [inline hack/install.sh:16]:
```suggestion
CGO_ENABLED=0 GOOS=$(go env GOOS) GOARCH=$(go env GOARCH) GO111MODULE=on \
```

File: pkg/deployer/container/init/init.go [hunk 2]
```diff
@@ -207,17 +206,28 @@ func (r registries) ArtifactsRegistry() artifactsregistry.Registry {
...
+			return nil
+		}
+
+		secrets = append(secrets, path)
+
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
```
Comment (schrodit) [inline pkg/deployer/container/init/init.go:227]:
```suggestion
		return nil, fmt.Errorf("unable to add local registry pull secrets: %w", err)
```

File: pkg/deployer/container/init/init.go [hunk 3]
```diff
@@ -207,17 +206,28 @@ func (r registries) ArtifactsRegistry() artifactsregistry.Registry {
...
 	}
 
-	keyring, err := credentials.CreateOCIRegistryKeyring(secrets, nil)
+	keyring, err := credentials.CreateOCIRegistryKeyringFromFilesyStem(nil, secrets, fs)
```
Comment (schrodit) [inline pkg/deployer/container/init/init.go:230]:
```suggestion
	keyring, err := credentials.CreateOCIRegistryKeyringFromFilesystem(nil, secrets, fs)
```

File: pkg/deployer/container/reconcile.go [hunk 4]
```diff
@@ -255,6 +278,164 @@ func setConditionsFromPod(pod *corev1.Pod, conditions []lsv1alpha1.Condition) []
...
+const (
+	dockerHubDomain       = "docker.io"
+	dockerHubLegacyDomain = "index.docker.io"
+)
+
+// syncSecrets identifies a authSecrets based on a given registry. It then creates or updates a pull secret in a target for the matched authSecret.
+func (c *Container) syncSecrets(ctx context.Context, secretName, registry string, authSecrets []*dockerconfigfile.ConfigFile) (string, error) {
+	imageref, err := dockerreference.ParseDockerRef(registry)
+	if err != nil {
+		return "", fmt.Errorf("Not a valid registry refrence %s: %w", c.ProviderConfiguration.Image, err)
```
Comment (schrodit) [inline pkg/deployer/container/reconcile.go:291]:
```suggestion
		return "", fmt.Errorf("Not a valid registry reference %s: %w", c.ProviderConfiguration.Image, err)
```

File: pkg/deployer/container/reconcile.go [hunk 5]
```diff
@@ -255,6 +278,164 @@ func setConditionsFromPod(pod *corev1.Pod, conditions []lsv1alpha1.Condition) []
...
+		if err != nil {
+			c.log.V(3).Info(fmt.Sprintf("Invalid auth config in secret %q, skipping", secretRef.NamespacedName().String()), "error", err.Error())
+			continue
+		}
+		secrets = append(secrets, authConfig)
+	}
+
+	var err error
+
+	// TODO: figure out what to do in case a private image on dockerhub is referenced but not with FQDN of registry (ie someone/something:latest)
```
Comment (schrodit) [inline pkg/deployer/container/reconcile.go:380]: is already figured out?

File: pkg/utils/oci/credentials/keyring.go [hunk 6]
```diff
@@ -29,8 +30,8 @@ type OCIKeyring interface {
 	Resolver(ctx context.Context, client *http.Client, plainHTTP bool) (remotes.Resolver, error)
 }
 
-// CreateOCIRegistryKeyring creates a new OCI registry keyring.
-func CreateOCIRegistryKeyring(pullSecrets []corev1.Secret, configFiles []string) (OCIKeyring, error) {
+// CreateOCIRegistryKeyringFromFileSytem creates a new OCI registry keyring from a given file system.
+func CreateOCIRegistryKeyringFromFilesyStem(pullSecrets []corev1.Secret, configFiles []string, fs vfs.FileSystem) (OCIKeyring, error) {
```
Comment (schrodit) [inline pkg/utils/oci/credentials/keyring.go:34]:
```suggestion
func CreateOCIRegistryKeyringFromFilesystem(pullSecrets []corev1.Secret, configFiles []string, fs vfs.FileSystem) (OCIKeyring, error) {
```

**Review by schrodit — CHANGES_REQUESTED** [review]

**Comment by schrodit** [comment]
Can you please rebase to also include changes that are done to the pipeline definition.
Then the publish step will eventually succeed.

File: pkg/deployer/container/reconcile.go (hunk 5 above)
Comment (hendrikKahl) [reply pkg/deployer/container/reconcile.go:384]: yes, I removed the comment

**Comment by hendrikKahl** [comment]
I squashed the commits to avoid issues while merging and included the suggestions in additon.

**Comment by hendrikKahl** [comment]
@MrBatschner fixed a few more things and we rebased to master.

File: pkg/deployer/container/actuator.go [hunk 7]
```diff
@@ -125,7 +116,13 @@ func (a *actuator) reconcile(ctx context.Context, deployItem *lsv1alpha1.DeployI
 			deployItem.Status.LastError,
 			5*time.Minute))
 	}()
-	containerOp, err := New(a.log, a.lsClient, a.hostClient, a.registry, a.config, deployItem)
+
+	// Todo sync pull secrets to container Namespace
```
Comment (schrodit) [inline pkg/deployer/container/actuator.go:120]: I think the secrets are correctly synced

File: pkg/deployer/container/reconcile.go [hunk 8]
```diff
@@ -255,6 +278,164 @@ func setConditionsFromPod(pod *corev1.Pod, conditions []lsv1alpha1.Condition) []
...
```
Comment (schrodit) [inline pkg/deployer/container/reconcile.go:380]: i think you missed that one

File: pkg/utils/oci/credentials/keyring.go [hunk 9]
```diff
@@ -29,8 +30,8 @@ type OCIKeyring interface {
 	Resolver(ctx context.Context, client *http.Client, plainHTTP bool) (remotes.Resolver, error)
 }
 
-// CreateOCIRegistryKeyring creates a new OCI registry keyring.
-func CreateOCIRegistryKeyring(pullSecrets []corev1.Secret, configFiles []string) (OCIKeyring, error) {
+// CreateOCIRegistryKeyringFromFileSytem creates a new OCI registry keyring from a given file system.
```
Comment (schrodit) [inline pkg/utils/oci/credentials/keyring.go:33]:
```suggestion
// CreateOCIRegistryKeyringFromFilesytem creates a new OCI registry keyring from a given file system.
```

**Review by schrodit — CHANGES_REQUESTED** [review]

**Review by schrodit — APPROVED** [review]
/lgtm

**Pull request merged**